libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/ApolloCall.kt

```diff
@@ -93,16 +106,34 @@ class ApolloCall<D : Operation.Data> internal constructor(
         .sendDocument(sendDocument)
         .enableAutoPersistedQueries(enableAutoPersistedQueries)
         .canBeBatched(canBeBatched)
+        .throwOnException(throwOnException)
         .build()
     return apolloClient.executeAsFlow(request)
   }
 
   /**
-   * A shorthand for `toFlow().single()`.
-   * Use this for queries and mutation to get a single [ApolloResponse] from the network or the cache.
-   * For subscriptions, you usually want to use [toFlow] instead to listen to all values.
+   * Retrieve a single [ApolloResponse] from this [ApolloCall], ignoring any cache misses or network errors.
+   *
+   * Use this for queries and mutations to get a single value from the network or the cache.
+   * For subscriptions or operations using `@defer`, you usually want to use [toFlow] instead to listen to all values.
+   *
+   * @throws ApolloException if the call returns zero or multiple valid GraphQL responses.
    */
   suspend fun execute(): ApolloResponse<D> {
-    return toFlow().single()
+    val responses = toFlow().toList()
+    val (errors, successes) = responses.partition { it.exception != null }
+    if (successes.size > 1) {
+      throw ApolloException("The operation returned multiple items, use .toFlow() instead of .execute()")
+    } else if (successes.isEmpty()) {
+      if (errors.size == 1) {
+        throw errors[0].exception!!
+      } else if (errors.size > 1) {
+        throw ApolloCompositeException(errors[0].exception, errors[1].exception)
+      } else {
+        throw ApolloException("The operation did not emit any item, check your interceptor chain")
+      }
```

Comment:
Should we honour `throwOnException` in `execute()` as well? Seems weird to have an API that requires catching exceptions and not the other one.

Reply:
Also reminder to update [the doc page for error handling](https://www.apollographql.com/docs/kotlin/essentials/errors)

Reply:
Yes, I was bothered by this because it needs to throw anyway here - but I agree it's a programming error, whereas for the regular case it makes sense to follow `throwOnException`. 

This change is more impactful I think (assuming `execute` is used a lot more than `toFlow`)! 
I don't think a migration helper could be made to automatically refactor code to be `if (response.exception != null)` instead of `try {} catch(e: ApolloException {}` (or maybe only for simple cases) but it could probably at least detect it and issue a warning.

I still think it's the right move: 
- it could be surprising to have `ApolloReponse.exception` but it being essentially useless because the exception is thrown instead
- generally speaking, throwing is not cool
- one can still set `throwOnException` to false (either globally on a specific call) (albeit it's deprecated)
- one can do `response.exception?.let { throw it }` or write an extension that does it

Reply:
I'm confused are you suggesting to keep as is and throw on network error (even if `throwOnException  == false`)?

Reply:
No, I'm agreeing we should throw if it's true, and not throw if it's false 😊
Also recognizing that since it impacts execute(), and it's false by default, it's somewhat a "bigger" change. Regardless, it makes sense to me.

Reply:
![giphy](https://user-images.githubusercontent.com/3974977/219666691-e7eb8e68-9d83-4cf4-ba33-db8dd90e7a64.gif)